pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -689,27 +698,67 @@ func CredentialHelpers(sys *types.SystemContext) ([]string, error) {
 	return config.partialV2.CredentialHelpers, nil
 }
 
-// refMatchesPrefix returns true iff ref,
+// refMatchingSubdomainPrefix returns the length of ref
+// iff ref, which is a registry, repository namespace, repository or image reference (as formatted by
+// reference.Domain(), reference.Named.Name() or reference.Reference.String()
+// — note that this requires the name to start with an explicit hostname!),
+// matches a Registry.Prefix value containing wildcarded subdomains in the
+// format: *.example.com. Wildcards are only accepted at the beginning, so
+// other formats like example.*.com will not work. Wildcarded prefixes also
+// cannot contain port numbers or namespaces in them.
+func refMatchingSubdomainPrefix(ref, prefix string) int {
+	index := strings.Index(ref, prefix[1:])
+	if index == -1 {
+		return -1
+	}
+	if strings.Contains(ref[:index], "/") {
+		return -1
+	}
+	index += len(prefix[1:])
+	if index == len(ref) {
+		return index
+	}
+	switch ref[index] {
+	case ':', '/', '@':
+		slashIndex := strings.LastIndex(ref, "/")
+		if slashIndex != -1 {
+			return slashIndex
```

Comment:
What does this do?

If I’m not mistaken, given prefix=`*.foo`, location=`destination/dest-ns`, and ref=`domain.foo/ns1/ns2/ns3/ns4`, this causes the rewrite to be `destination/dest-ns/ns4`.

Reply:
> What does this do?
> 
> If I’m not mistaken, given prefix=`*.foo`, location=`destination/dest-ns`, and ref=`domain.foo/ns1/ns2/ns3/ns4`, this causes the rewrite to be `destination/dest-ns/ns4`.

Sorry I didn't get the clarified early on, what should the expected rewrite be in the above case?

Would it be `destination/dest-ns/ns4` OR `destination/dest-ns/ns1/ns2/ns3/ns4` OR something else?

Reply:
i'll add a check for such a case too.

Reply:
Most importantly, it should be consistent with a non-wildcard prefix.

I’d expect `destination/dest-ns/ns1/ns2/ns3/ns4`: the match is rewritten, nothing else is affected.

Reply:
ok, I will add that in, along with a test case, thanks a lot!!

Reply:
addressed in latest version.